src/lib/logstore/log_dev.cpp

```diff
@@ -580,7 +583,242 @@ void LogDev::rollback(logstore_id_t store_id, logid_range_t id_range) {
+    const auto it = m_id_logstore_map.find(store_id);
+    HS_REL_ASSERT((it == m_id_logstore_map.end()), "store_id {}-{} already exists", m_logdev_id, store_id);
+    LOGINFO("Opening log store id {}-{}", m_logdev_id, store_id);
+    m_id_logstore_map.insert(std::make_pair<>(store_id, logstore_info_t{nullptr, on_open_cb, append_mode}));
+}
+
+void LogDev::remove_log_store(logstore_id_t store_id) {
+    LOGINFO("Removing log store id {}-{}", m_logdev_id, store_id);
+
+    {
+        folly::SharedMutexWritePriority::WriteHolder holder(m_store_map_mtx);
+        auto ret = m_id_logstore_map.erase(store_id);
+        HS_REL_ASSERT((ret == 1), "try to remove invalid store_id {}-{}", m_logdev_id, store_id);
+    }
+    unreserve_store_id(store_id);
+}
+
+void LogDev::device_truncate_under_lock(const std::shared_ptr< truncate_req >& treq) {
+    run_under_flush_lock([this, treq]() {
+        iomanager.run_on_forget(logstore_service().truncate_thread(), [this, treq]() {
+            const logdev_key trunc_upto = do_device_truncate(treq->dry_run);
+            bool done{false};
+            if (treq->cb || treq->wait_till_done) {
+                {
+                    std::lock_guard< std::mutex > lk{treq->mtx};
+                    done = (--treq->trunc_outstanding == 0);
+                    treq->m_trunc_upto_result[m_logdev_id] = trunc_upto;
+                }
+            }
+            if (done) {
+                if (treq->cb) { treq->cb(treq->m_trunc_upto_result); }
+                if (treq->wait_till_done) { treq->cv.notify_one(); }
+            }
+            unlock_flush();
+        });
+        return false; // Do not release the flush lock yet, the scheduler will unlock it.
+    });
+}
+
+void LogDev::on_log_store_found(logstore_id_t store_id, const logstore_superblk& sb) {
+    folly::SharedMutexWritePriority::ReadHolder holder(m_store_map_mtx);
+    const auto it = m_id_logstore_map.find(store_id);
+    if (it == m_id_logstore_map.end()) {
+        LOGERROR("Store Id {}-{} found but not opened yet.", m_logdev_id, store_id);
+        m_unopened_store_id.insert(store_id);
+        m_unopened_store_io.insert(std::make_pair<>(store_id, 0));
+        return;
+    }
+
+    LOGINFO("Found a logstore store_id={}-{} with start seq_num={}, Creating a new HomeLogStore instance", m_logdev_id,
+            store_id, sb.m_first_seq_num);
+    auto& l_info = const_cast< logstore_info_t& >(it->second);
+    l_info.m_log_store =
+        std::make_shared< HomeLogStore >(shared_from_this(), store_id, l_info.append_mode, sb.m_first_seq_num);
+    if (l_info.m_on_log_store_opened) l_info.m_on_log_store_opened(l_info.m_log_store);
+}
+
+static thread_local std::vector< std::shared_ptr< HomeLogStore > > s_cur_flush_batch_stores;
+
+void LogDev::on_io_completion(logstore_id_t id, logdev_key ld_key, logdev_key flush_ld_key,
+                              uint32_t nremaining_in_batch, void* ctx) {
```

Comment:
Suggestion: In fact this whole function could be moved under on_flush_completion().

Reply:
The same callback is overridden by test code test_log_dev